File: src/my_cpp_pkg/src/led_panel.cpp
#include <memory>
#include <functional>
#include <chrono>
#include <array>

#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/led_panel_state.hpp"
#include "my_robot_interfaces/srv/set_led_state.hpp"

using namespace std::placeholders;

class LedPanelNode : public rclcpp::Node
{
public:
    LedPanelNode() : Node("led_panel")
    {
        panel_state_publisher_ = create_publisher<my_robot_interfaces::msg::LedPanelState>(
            "led_panel_state", 10);
        panel_state_publishing_timer_ = create_wall_timer(
            std::chrono::seconds(1), std::bind(&LedPanelNode::publishPanelState, this));
        set_led_service_ = create_service<my_robot_interfaces::srv::SetLedState>(
            "set_led",
            std::bind(&LedPanelNode::callbackSetLed, this, _1, _2));
        RCLCPP_INFO(get_logger(), "Led Panel has been started.");
    }

private:
    void publishPanelState()
    {
        auto msg = my_robot_interfaces::msg::LedPanelState();
        msg.state = panel_state_;
        panel_state_publisher_->publish(msg);
    }

    void callbackSetLed(const my_robot_interfaces::srv::SetLedState::Request::SharedPtr request,
                          my_robot_interfaces::srv::SetLedState::Response::SharedPtr response)
    {
        if (request->led_number > 0 && request->led_number < 4)
        {
            panel_state_[request->led_number - 1] = request->state;
            response->success = true;
            publishPanelState();
            RCLCPP_INFO(get_logger(), "Updated LED %d to %s", request->led_number, request->state ? "true" : "false");
        }
        else
        {
            response->success = false;
            RCLCPP_WARN(get_logger(), "Requested LED number must be 1, 2, or 3.");
        }
    }

    std::array<bool, 3> panel_state_;
    rclcpp::Publisher<my_robot_interfaces::msg::LedPanelState>::SharedPtr panel_state_publisher_;
    rclcpp::Service<my_robot_interfaces::srv::SetLedState>::SharedPtr set_led_service_;
    rclcpp::TimerBase::SharedPtr panel_state_publishing_timer_;
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<LedPanelNode>());
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/led_panel.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import LedPanelState
from my_robot_interfaces.srv import SetLedState

class LedPanelNode(Node):

    def __init__(self):
        super().__init__("led_panel")
        self.panel_state_ = [False, False, False]
        self.panel_state_publisher_ = self.create_publisher(
            LedPanelState, "led_panel_state", 10)
        self.panel_state_publishing_timer_ = self.create_timer(1.0, self.publish_panel_state)
        self.set_led_service_ = self.create_service(
            SetLedState, "set_led", self.callback_set_led)
        self.get_logger().info("Led Panel has been started.")

    def publish_panel_state(self):
        msg = LedPanelState()
        msg.state = self.panel_state_
        self.panel_state_publisher_.publish(msg)

    def callback_set_led(self, request, response):
        led_number = request.led_number
        state = request.state

        if led_number > len(self.panel_state_) or led_number <= 0:
            response.success = False
            return response

        self.panel_state_[led_number - 1] = state
        response.success = True
        self.publish_panel_state()
        return response


def main(args=None):
    rclpy.init(args=args)
    node = LedPanelNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
